ratis-server/src/main/java/org/apache/ratis/server/impl/ServerState.java

```diff
@@ -104,17 +104,24 @@
     configurationManager = new ConfigurationManager(initialConf);
     LOG.info("{}: {}", getMemberId(), configurationManager);
 
+    boolean storageFound = false;
     List<File> directories = RaftServerConfigKeys.storageDir(prop);
     while (!directories.isEmpty()) {
       // use full uuid string to create a subdirectory
       File dir = chooseStorageDir(directories, group.getGroupId().getUuid().toString());
       try {
         storage = new RaftStorageImpl(dir, RaftServerConfigKeys.Log.corruptionPolicy(prop));
+        storageFound = true;
         break;
-      } catch (AccessDeniedException e) {
-        directories.remove(dir);
+      } catch (IOException e) {
+        directories.removeIf(d -> d.getAbsolutePath().equals(dir.getParent()));
```

Comment:
If we don't remove it, the while-loop won't terminate.

Reply:
@szetszwo here is the BUG, in the original code, `dir` represents the newly created directory(name = UUID) for the group under the storageDir, not the storageDir itself, so it loops forever indeed under my test.
So here I try to remove the exact storageDir, do you have a better way to write this line in Java?

Reply:
@szetszwo because the original `remove(dir)` is a BUG I think.
Here dir points to '/data1/hdds-ratis/<groupID>', but I believe we want to iterator through all the storageDirs(/data1/hdds-ratis,/data2/hdds-ratis,...), and remove some 'dir' that is "BAD" which should points to '/dataX/hdds-ratis', not '/data1/hdds-ratis/<groupID>'.
Then I use removeIf() to filter through the storageDirs and remove the exact storageDir that is BAD, so `dir.getParent()` here.

Reply:
I see.  That's a good point.  Thanks.